ml4h/applications/ingest/ingest_xml_metadata.py

```diff
@@ -0,0 +1,318 @@
+import xml.etree.ElementTree as ET
+import string
+import uuid
+
+# Set of printable characters.
+printable = set(string.printable)
+
+# Constant: ignore these keys in the XML during ingestion.
+# For the UK Biobank ECG XMLs:
+ukb_ignore_elements = [
+    "Stripdata",
+    "Fulldisclosure",
+    "Mediansamples",
+    "Waveformdata",
+    "Fulldisclosuredata",
+]
+# For Partners (MUSE) ECG XMLs:
+muse_ignore_elements = ['MeasurementMatrix','WaveFormData']
+
+# Magic
+HANDLERS = {
+    "property": "handle_property",
+}
+
+# For the UKB ECG XMLs
+def _sample_id_from_path(path):
+    # Assumes thef format UKBID_FIELDID_INSTANCE.csv
+    base = os.path.split(path)[-1].split(".")[0].split("_")
+    return base[0], base[2]  # ID, INSTANCE
+
+
+# For MUSE ECG XMLs
+# def _sample_id_from_path(path):
+#     # base = ?
+#     return '%032x' % random.randrange(16**32),0 # random, nonsense
+
+
+def generate_output_name(
+    subject_id: str, instance: str, source: str = "ukb", data_type: str = "ecg"
+) -> str:
+    """Helper function to generate output names. Current implementation will
+    return `BROAD_ML4H_mdrk_{source}_{data_type}_metadata_{id}_i{instance}_{hash}`.
+
+    Examples:
+        * BROAD_ML4H_mdrk_ukb_ecg_metadata_6025769_i2_cc96e96cbe03272703223adf
+        * BROAD_ML4H_mdrk_muse_ecg_metadata_2331765_i0_545d840638e5fe00592ee64d.pq
+
+    Args:
+        subject_id (str): MUSE or UKB subject identifier. This could also be any random value for privacy.
+        instance (str): UKB instance number. For MUSE this can be anything.
+        source (str, optional): XML source: currently `ukb` or `muse` but no restrictions apply. Defaults to 'ukb'.
+        data_type (str, optional): Modality associated with the XML. Defaults to 'ecg'.
+
+    Returns:
+        str: A prepared output string name.
+    """
+    return f"BROAD_ML4H_mdrk_{str(source)}_{data_type}_metadata_{str(subject_id)}_i{str(instance)}_{uuid.uuid4().hex}"
+
+
+def _recurseTree(root, path: str, store_dict: dict, ignore_elements: list):
+    """Recursive function for iterating over the XML tree and extracting out all
+    keys and values for those keys that are not in the `ignore_elements` black list.
+    Note: All values will be stored as strings and all strings are scrubbed for
+    newlines, tabs, and trailing and leading white space. Multiple elements for the
+    same putative path will be stored as an array of the values for the same key.
+    Nones are stored as empty strings ('').
+
+    Args:
+        root: Current element in the XML tree.
+        path (str): Current string name for the traversed path.
+        store_dict (dict): Output dictionary.
+        ignore_elements (list): List of keys to ignore
+    """
+    if root.tag in ignore_elements or root.tag.title() in ignore_elements:
+        return
+
+    global HANDLERS  # Not required for this application
+
+    if root.tag in HANDLERS:
+        handler = globals()[HANDLERS[root.tag]]
+        handler(root)
```

Comment:
can you explain what this does?

Reply:
This does nothing for this application. Removed